### PR TITLE
feat: light SDK generation

### PR DESCRIPTION
**Pull request by sdo-1A**
## Proposed change

Possibility of generating typescript SDK without revivers when they are not necessary.

## Related issues

- :rocket: Feature [#(1337)](https://github.com/AmadeusITGroup/otter/issues/1337)

### CODE REVIEW

File: packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/AbstractTypeScriptClientCodegen.java [hunk 1]
```diff
@@ -716,6 +717,23 @@ public Map<String, ModelsMap> postProcessAllModels(Map<String, ModelsMap> objs)
...
+    Predicate<ModelMap> isNotEnumPredicate = modelMap -> !modelMap.getModel().isEnum;
+    Predicate<ModelMap> requiredDictionaryPredicate = modelMap -> Boolean.TRUE.equals(modelMap.getModel().vendorExtensions.get("requireDictionary"));
+    Predicate<ModelMap> anyVarRequiredReviverPredicate = modelMap -> modelMap.getModel().allVars.stream().anyMatch(varRequiredReviverPredicate);
+    boolean reviversRequired = objs.values().stream().anyMatch(
+      modelsMap -> modelsMap.getModels().stream().anyMatch(
+        isNotEnumPredicate.and(requiredDictionaryPredicate.or(anyVarRequiredReviverPredicate))
+      )
+    );
+    // Setting keepRevivers value in each model since the additionalProperties are set up before the post process
+    for (Map.Entry<String, ModelsMap> entry : objs.entrySet()) {
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/AbstractTypeScriptClientCodegen.java:730]:
```java
objs.values().stream().forEach(modelsMap -> modelsMap.put("keepRevivers", reviversRequired))
```
It may be enough

**Comment by kpanot** [comment]
You should add documentation regarding the enabling/disabling process to the light SDK feature

**Comment by matthieu-crouzet** [comment]
> You should add documentation regarding the enabling/disabling process to the light SDK feature
It will be done in another PR
https://github.com/AmadeusITGroup/otter/issues/1338

File: packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/resources/typescriptFetch/api/api.mustache [hunk 2]
```diff
@@ -1,7 +1,9 @@
 {{#noUnusedImport}}
 {{#imports}}
 import { {{import}} } from '../../models/base/{{#kebabCase}}{{import}}{{/kebabCase}}/index';
+{{#keepRevivers}}
```
Comment (kpanot) [inline packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/resources/typescriptFetch/api/api.mustache:4]: You may need to document the behavior (not generating the reviver per default).

File: packages/@ama-sdk/schematics/README.md [hunk 3]
```diff
@@ -79,12 +79,28 @@ If you use `Yarn2+`, you can use the following `scripts` in `package.json`:
 
 Use `generate` to (re)generate your SDK based on the content of `./swagger-spec.yaml` (make sure you have this file at the root of your project) and `upgrade:repository` to regenerate the structure of your project.
 
+#### Light SDK
+
+For SDKs that do not need to handle extensibility, the use of dictionaries or dates, a "light" version of the SDK will be generated.
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/schematics/README.md:84]: Light sdk is not directly link to dates but more to revivers 
because we could have date but in string

File: packages/@ama-sdk/schematics/README.md [hunk 4]
```diff
@@ -79,12 +79,28 @@ If you use `Yarn2+`, you can use the following `scripts` in `package.json`:
 
 Use `generate` to (re)generate your SDK based on the content of `./swagger-spec.yaml` (make sure you have this file at the root of your project) and `upgrade:repository` to regenerate the structure of your project.
 
+#### Light SDK
+
+For SDKs that do not need to handle extensibility, the use of dictionaries or dates, a "light" version of the SDK will be generated. 
+These SDKs do not need to generate revivers, which results in a simplified version of the code and reduces the size of the bundle.
+
+If your specification file includes dates, there are multiple options for the generation of your SDK involving the global property option `stringifyDate`: 
+* By default, the option `stringifyDate` is false so dates will be generated as expected in the specifications.
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/schematics/README.md:88]:
```suggestion
* By default, the option `stringifyDate` is false so dates will be generated as utils.Date/utils.DateTime/Date
```

File: packages/@ama-sdk/schematics/README.md [hunk 5]
```diff
@@ -79,12 +79,28 @@ If you use `Yarn2+`, you can use the following `scripts` in `package.json`:
...
+* If the dates can be of type `string`, setting the option `stringifyDate` to true will ensure the generation of a light SDK. <br>
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/schematics/README.md:89]: Not really true 
we can set `stringifyDate` and still have revivers because of `dictionaries`

Comment (cpaulve-1A) [inline packages/@ama-sdk/schematics/README.md:89]: Also the "can be of type 'string'" might be a bit confusing for the users. Be more direct in your approach and be more explicite.
For example:
To keep the handling of timezone to the application and handle date as string, the option stringifyDate should be set to false

File: packages/@ama-sdk/schematics/README.md [hunk 6]
```diff
@@ -79,12 +79,28 @@ If you use `Yarn2+`, you can use the following `scripts` in `package.json`:
...
+  This can be done by adding `--global-property stringifyDate` to the generator command.
```
Comment (cpaulve-1A) [inline packages/@ama-sdk/schematics/README.md:90]: Is it possible to set it in the openapi json configuration file?

File: packages/@ama-sdk/showcase-sdk/src/api/store/store-api.ts [hunk 7]
```diff
@@ -105,7 +104,7 @@ export class StoreApi implements Api {
     const options = await this.client.prepareOptions(basePathUrl, 'GET', getParams, headers, body || undefined, tokenizedOptions, metadata);
     const url = this.client.prepareUrl(options.basePath, options.queryParams);
 
-    const ret = this.client.processCall<{ [key: string]: number }>(url, options, ApiTypes.DEFAULT, StoreApi.apiName, { 200: undefined } , 'getInventory');
+    const ret = this.client.processCall<{ [key: string]: number }>(url, options, ApiTypes.DEFAULT, StoreApi.apiName, undefined, 'getInventory');
```
Comment (cpaulve-1A) [inline packages/@ama-sdk/showcase-sdk/src/api/store/store-api.ts:107]: why this change?

Comment (sdo-1A) [reply packages/@ama-sdk/showcase-sdk/src/api/store/store-api.ts:107]: since this is the `revivers?` parameter of the `processCall` function

File: packages/@ama-sdk/schematics/README.md [hunk 8]
```diff
@@ -79,12 +79,45 @@ If you use `Yarn2+`, you can use the following `scripts` in `package.json`:
...
+It is also possible to generate a "light" SDK. This type of SDK does not need to generate revivers, which results in a simplified version of the code and reduces the size of the bundle.
+
+Please note that revivers are generated for SDKs that use:
+* non-stringified dates
+* dictionaries
+* extensible models
+
+##### Dates
+If your specification file includes dates, there are multiple options for the generation of your SDK involving the global property option `stringifyDate`: 
+* By default, the option `stringifyDate` is false so dates will be generated as either `utils.Date`, `utils.DateTime`, or `Date`.
```
Comment (mrednic-1A) [inline packages/@ama-sdk/schematics/README.md:93]: How is done the difference when generating a javascript `Date` or an `utils.Date`. Is it another option to be passed to the generator ?

File: packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/AbstractTypeScriptClientCodegen.java [hunk 9]
```diff
@@ -534,8 +542,8 @@ public void postProcessModelProperty(CodegenModel model, CodegenProperty propert
         + "list container.");
     }
 
-    // If the x-date-timezone is present in the specification, replace Date with utils.Date
-    if (property.vendorExtensions.containsKey("x-date-timezone")) {
+    // If the x-date-timezone is present in the specification, replace utils.Date with Date
+    if ("utils.Date".equals(property.dataType) && property.vendorExtensions.containsKey("x-date-timezone")) {
```
Comment (fpaul-1A) [inline packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/AbstractTypeScriptClientCodegen.java:546]: shouldn't we check `utils.DateTime` as well ?

**Review by vscaiceanu-1a — DISMISSED** [review]

File: packages/@ama-sdk/schematics/README.md (hunk 8 above)
Comment (sdo-1A) [reply packages/@ama-sdk/schematics/README.md:93]: I believe this is explained in the doc that I linked in the line below, the difference depends on the x-date-timezone vendor in the specs (https://github.com/AmadeusITGroup/otter/tree/main/packages/%40ama-sdk/schematics/schematics/typescript/shell/templates/base#manage-dates)

File: packages/@ama-sdk/schematics/README.md (hunk 6 above)
Comment (sdo-1A) [reply packages/@ama-sdk/schematics/README.md:96]: not yet, this will be done in another PR

File: packages/@ama-sdk/schematics/README.md (hunk 8 above)
Comment (mrednic-1A) [reply packages/@ama-sdk/schematics/README.md:93]: ok I see :) I missed that info.

**Review by fpaul-1A — APPROVED** [review]

**Review by vscaiceanu-1a — APPROVED** [review]

**Review by cpaulve-1A — APPROVED** [review]
